streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -102,14 +127,98 @@ public void process(final K key, final V1 value) {
+                        outerJoinWindowStore.ifPresent(store -> store.put(
+                            KeyAndJoinSide.make(thisJoin, key),
+                            makeValueOrOtherValue(thisJoin, value),
+                            inputRecordTimestamp));
+                    }
+                }
+
+                outerJoinWindowStore.ifPresent(store -> {
+                    // only emit left/outer non-joined if the stream time has advanced (inputRecordTime = maxStreamTime)
+                    // if the current record is late, then there is no need to check for expired records
+                    if (inputRecordTimestamp == maxStreamTime) {
+                        maybeEmitOuterExpiryRecords(store, maxStreamTime);
+                    }
+                });
+            }
+        }
+
+        private ValueOrOtherValue makeValueOrOtherValue(final boolean thisJoin, final V1 value) {
+            return thisJoin
+                ? ValueOrOtherValue.makeValue(value)
+                : ValueOrOtherValue.makeOtherValue(value);
+        }
+
+        @SuppressWarnings("unchecked")
+        private void maybeEmitOuterExpiryRecords(final WindowStore<KeyAndJoinSide<K>, ValueOrOtherValue> store, final long maxStreamTime) {
+            try (final KeyValueIterator<Windowed<KeyAndJoinSide<K>>, ValueOrOtherValue> it = store.all()) {
+                while (it.hasNext()) {
+                    final KeyValue<Windowed<KeyAndJoinSide<K>>, ValueOrOtherValue> e = it.next();
+
+                    // Skip next records if the oldest record has not expired yet
+                    if (e.key.window().end() + joinGraceMs >= maxStreamTime) {
+                        break;
+                    }
+
+                    final K key = e.key.key().getKey();
+
+                    // Emit the record by joining with a null value. But the order varies depending whether
+                    // this join is using a reverse joiner or not. Also whether the returned record from the
+                    // outer window store is a V1 or V2 value.
+                    if (thisJoin) {
+                        if (e.key.key().isThisJoin()) {
+                            context().forward(key, joiner.apply(key, (V1) e.value.getThisValue(), null));
+                        } else {
+                            context().forward(key, joiner.apply(key, null, (V2) e.value.getOtherValue()));
+                        }
+                    } else {
+                        if (e.key.key().isThisJoin()) {
+                            context().forward(key, joiner.apply(key, null, (V2) e.value.getThisValue()));
+                        } else {
+                            context().forward(key, joiner.apply(key, (V1) e.value.getOtherValue(), null));
+                        }
+                    }
+
+                    // Delete the key from tne outer window store now it is emitted
+                    store.put(e.key.key(), null, e.key.window().start());
```

Comment:
This reminds me about the test coverage: maybe we should also test that store.put / delete can be triggered while the iterator is open, and if the put / deleted elements would not be reflected from the iterator.

Reply:
I'm not sure if this is possible. If I put or delete a future record that the iterator hasn't seen, then the iterator will reflect that change if I continue iterating until the deleted or new record appears. This is how Rocksdb iterator works, isn't it? When asking the next record, it only reads one record from the store. If subsequent records are deleted or are new, then the iterator will read the update in the next next() call.

Or what test case do you have in mind?

Reply:
Sorry it's my bad, I mis read the code and thought you were deleting records while inside the

```
try (final KeyValueIterator<Windowed<KeyAndJoinSide<K>>, LeftOrRightValue> it = store.all()) {
```

block. Please ignore this comment.